.github/setup-node/action.yml

```diff
@@ -0,0 +1,58 @@
+name: 'Setup Node'
+description: 'Setup node and install dependencies with caching support'
+inputs:
+    node-version:
+        description: 'Optional node version to pass to setup-node'
+        required: false
+        type: string
+
+    should-rebuild:
+        description: 'Whether to rebuild the dependencies after --ignore-scripts'
+        default: false
+        required: false
+        type: boolean
+
+runs:
+    using: 'composite'
+    steps:
+        - name: Use desired version of NodeJS from input
+          if: ${{ inputs.node-version != '' }}
+          uses: actions/setup-node@v3
+          with:
+              node-version: ${{ inputs.node-version }}
+              cache: npm
+
+        - name: Use desired version of NodeJS from .nvmrc
+          if: ${{ inputs.node-version == '' }}
+          uses: actions/setup-node@v3
+          with:
+              node-version-file: '.nvmrc'
+              cache: npm
+
+        - name: Get the node version
+          id: node-version
+          run: echo "NODE_VERSION=$(node -v)" >> $GITHUB_OUTPUT
+          shell: bash
+
+        - name: Cache node_modules
+          id: cache-node_modules
+          uses: actions/cache@v3
+          with:
+              path: '**/node_modules'
+              key: node_modules-${{ runner.os }}-${{ steps.node-version.outputs.NODE_VERSION }}-${{ hashFiles('package-lock.json') }}
```

Comment:
[`restore-keys`](https://docs.github.com/en/actions/using-workflows/caching-dependencies-to-speed-up-workflows#matching-a-cache-key) is not needed because we'll be running `npm ci` if there isn't an exact match anyway. `npm ci` will remove all of the `node_modules`.

Reply:
do we need to include the `npm` version in here? I know we already have issues where newer `npm` won't work because of conflicts with package resolution between the versions, or in the `package-lock.json` format.

Reply:
right, but if we update the `npm` version and use the cache from the old one then won't that lead to a mismatch in the installed packages?

Reply:
I'm not sure. I don't know if anything will change except for the built dependencies. It seems like if I compile the existing project with different versions of `npm` I _have_ to clear out `node_modules` or `npm ci` fails, and possibly won't run with a mix of modules built in the old version and some built in the new. I'm just double-checking here because this introduces what could be a opportunity for things to break on account of this, and again, in a very nebulous manner that won't appear obvious in the logs.

you are probably right; but I'd want us to be very sure of that

Reply:
I don't mind including it in the cache key as it also rarely changes. Appreciate the effort for trying to keep it safe!